### DISCUSSION TITLE
Not charging at the lowest possible price

### DISCUSSION THREAD

**Discussion by jan-happy** [sonnen-tools/node-red#18] [Q&A] [answered]
Hi,
first of all: Thank you so much for your work. I'm totally new to Node Red and your Flows seem to work (almost ;) perfectly out of the box.
So maybe this issue I'm reporting is just related to rtfm, but I could not figure out why the flow would start charing at 21:00 when the price was at 28 Cent? At 2:00 it would only have cost 25 Cent.
![IMG_8248](https://github.com/sonnen-tools/node-red/assets/695826/6a8bff63-62d6-4955-97bf-0d9b2f0a64ca)
![IMG_8249](https://github.com/sonnen-tools/node-red/assets/695826/44641274-f7f9-4869-a7e5-53bda5754ab6)

**Comment by dlindh** [comment]
Hi :) 

I am not the developer  just a user :)

What is your maxPrice in Cheapest Hours Config? That will decide the cheapest price you charge at? (Together with conversion_loss_factor)
 
https://github.com/sonnen-tools/node-red/blob/main/docs/configuration.md

**Reply by jan-happy** [reply]
Thanks! It's currently set to 0.31
But I thought that the algorithm would look at all the prices in the future and start loading at the cheapest possible rate!?

**Reply by markusjnagel** [reply]
Thanks, Daniel @dlindh for chiming in :-) - much appreciated :+1: 
And, hi @jan-happy - glad this is of some use to you 

If you could go to http://_yourrpi-address_:1880/status and attach a screenshot of that, please? 

But, I agree - that is not ideal. It works as follows:

It calculates the average over 
- the whole Tibber time period available (setting: `average_all`)
- the most expensive five hours in the time period available (setting: `average_five`)

From that average, it the calculates the maximum price you could charge for without losing money (aka it factors in your conversion loss setting).

That gives you the maximum price where it still makes sense to charge. At times during high prices, that might be higher than what you are willing to pay in general - that is what the User maxPrice is for (it will never charge for a higher price, regardless what the calculation says).

### And now the conclusion     :drum: _drumroll_ :drum: 

All hours that are identified as "cheaper than what I just calculated" will be eligible for charging and saved as "cheap Hours". 
The algorithm (that's actually too posh a word for this simple thingy) will simply allow charging for all hours that are "cheap Hours" - even if there is an even cheaper hour coming up.... 
Yeah, not ideal.... 

I'm totally open to suggestions (I'm super selfish here, so if someone can describe the golden rule in a way that I can program it - I would benefit as well :smile: )

### Alternatives already under construction 
@dlindh has a use case (see [this discussion](https://github.com/orgs/sonnen-tools/discussions/7)) where he would benefit from discharging only during the x hours with the highest price (otherwise his battery is empty before the really expensive hours hit).
I'm currently building that. 

Similarly, I could build a simple rule that says: Regardless of averages, just pick the cheapest x hours of the day and charge (unless the price is higher than my User maxPrice). 

What do you think?

**Reply by dlindh** [reply]
Would it be possible to solve like this?
A setting for battery size.
A settings for what is the lowest price you want to discharge for.
Existing setting for charge watt.
Existing setting for maxprice.

So in my case with a 20kWh battery if it where empty it would take around 4 hours to fill at around 5kW charge power.

So looking at the lowest priced hours  calculate the best hours to charge to fill the battery before the price has increased enough for the discharge price to be enough.

Would that work?

In this case maxprice could really be set high since the discharge always will give a nice payout. This might only be for northen europe though since our prices can differ  quite alot over the day.

//Daniel

**Reply by markusjnagel** [reply]
> his might only be for northen europe though since our prices can differ quite alot over the day. 

I think that would benefit everyone with
- larger differences over the day
- high consumption (where you would miss the most expensive hours because your battery is empty before you would benefit the most).

Good idea. :+1:

**Comment by doncuco** [comment]
We ha an other situation in Germany yesterday that we should consider to think about.

The days before the prices were very high with 36cent in max and minimums of 28cent for a few hours only.
Then one day there was a window where my battery found a slot to load. An did that with a discharging price of 34cent.
So the battery could come to a state where it would not discharge, because the prices dropped. A made a discussion for that.

But an other thing is, that we had constant low prices yesterday and we should have used them to load.

So if you would log the charging or discharging costs of the last month and would compare those with the expected prices with loses for yesterday, you could send the battery back to charging because this is overall a good situation.

What do you think?

**Reply by markusjnagel** [reply, marked as answer]
Yeah, and (no kidding) with some data we could train a machine learning model - which is part of what I do as a day job. 

However, we have to deal with the limited capacity of a Raspberry Pi...
Not talking about training a model (that would have to be done outside of a Pi anyway, as it requires computing power), but aside from that I would have to test the limits of data storage.... if we build a month-long log, Node-RED can persist that on disk, but it would be still loaded into memory to work with.

So, I'd like to make sure that we have as many features as possible that would still run on a teeny-tiny [Pi Zero 2W](https://github.com/sonnen-tools/node-red/blob/main/docs/setup/pi/install-pi.md#my-pi) - to keep the barrier for new users as low as possible. 
However, as I said - let me test that and see how much that thing can handle... and then we can maybe check and decide together which features would require a bigger Pi or additional / external hardware. 
There is, for example the [Google Coral](https://coral.ai/products/accelerator/) that makes inferencing* possible on small devices...   training a model is a different thing, though... 

*Inferencing: is the process of using a trained model to make predictions or decisions on new, unseen data. It is the stage where the model applies its learned knowledge to input data and produces an output.
